node/src/components/chain_synchronizer/operations.rs

```diff
@@ -68,54 +69,147 @@ impl<'a> Drop for ScopeTimer<'a> {
+            // Note: This is currently quadratic in the amount of peers (e.g. in a network with 400
+            // out of 401 bad peers, this would result in 160k comparisons), but we estimate that
+            // this is fine given the expected low number of bad peers for now. If this proves a
+            // problem, proper sets should be used instead.
+            //
+            // Using a vec is currently convenient because it allows for FIFO-ordered redemption.
+            peers.retain(|p| !bad_peer_list.contains(p));
+        }
+
+        let redemption_interval = self.config.redemption_interval as i64;
+        if redemption_interval != 0
+            && self.filter_count.fetch_add(1, Ordering::Relaxed) > redemption_interval
+        {
+            self.filter_count
+                .fetch_sub(redemption_interval, Ordering::Relaxed);
+
+            // Redeem the oldest bad node.
+            self.bad_peer_list
+                .write()
+                .expect("bad peer list lock poisoned")
+                .pop_front();
+        }
+    }
+
+    /// Marks a peer as bad.
+    fn mark_bad_peer(&self, peer: NodeId) {
+        let mut bad_peer_list = self
+            .bad_peer_list
+            .write()
+            .expect("bad peer list lock poisoned");
+
+        // Note: Like `filter_bad_peers`, this may need to be migrated to use sets instead.
+        if bad_peer_list.contains(&peer) {
+            bad_peer_list.push_back(peer);
+            info!(%peer, "peer already marked as bad for syncing");
+        } else {
+            info!(%peer, "marked peer as bad for syncing");
+        }
+    }
+
+    /// Clears the list of bad peers.
+    fn redeem_all(&self) {
+        let mut bad_peer_list = self
+            .bad_peer_list
+            .write()
+            .expect("bad peer list lock poisoned");
+
+        let bad_peer_list = mem::take::<VecDeque<NodeId>>(&mut bad_peer_list);
+        if !bad_peer_list.is_empty() {
+            warn!(?bad_peer_list, "ran out of peers, redeemed all bad peers")
+        }
+    }
 }
 
 /// Restrict the fan-out for a trie being retrieved by chunks to query at most 10 peers at a time.
 const TRIE_CHUNK_FETCH_FAN_OUT: usize = 10;
 
 /// Fetches an item. Keeps retrying to fetch until it is successful. Not suited to fetching a block
 /// header or block by height, which require verification with finality signatures.
-async fn fetch_retry_forever<T>(
-    effect_builder: EffectBuilder<JoinerEvent>,
-    config: &Config,
-    id: T::Id,
-) -> FetchResult<T>
+async fn fetch_retry_forever<T>(ctx: &ChainSyncContext<'_>, id: T::Id) -> FetchResult<T>
 where
     T: Item + 'static,
     JoinerEvent: From<FetcherRequest<T>>,
 {
     loop {
-        for peer in effect_builder.get_fully_connected_peers().await {
+        let mut new_peer_list = ctx.effect_builder.get_fully_connected_peers().await;
+        ctx.filter_bad_peers(&mut new_peer_list);
+
+        for peer in new_peer_list {
```

Comment:
`new_peer_list` may contain bad nodes. We should get the peers from `ctx`.

Reply:
It's supposed to be this way around though. `new_peer_list` is all nodes, while we only track the bad nodes in `ctx`. Then we use the list of bad nodes in `ctx` to remove those from `new_peer_list` and are left with only the non-bad aka good ones.

Reply:
I see. I hate mutating a value inside a function. Can we consume `new_peer_list` and return filtered out collection?

Reply:
That'd needless copy the entire thing and reallocate though, whereas this allows `retain` to save the allocation (I hope). Do you think that's worth it?